### DISCUSSION TITLE
wxpython embedding?

### DISCUSSION THREAD

**Discussion by domagojhack** [raphaelquast/EOmaps#82] [Q&A] [answered]
Just a question. As EOmaps is based on matplotlib for plotting, is it possible to use it for embedding maps into wxpython based app? Matplotlib plot can be embedded so I imagine EOmaps could be embedded too.

Cheers.

**Comment by raphaelquast** [comment]
Hey, interesting question, thanks for bringing it up!
So far, I don't have any experience with wxpython, but if it is capable of embedding matplotlib figures the static plotting capabilities of EOmaps should definitely work right away. 

For the dynamic part (e.g. annotations overlays etc.), EOmaps uses a lot of [blitting](https://matplotlib.org/stable/tutorials/advanced/blitting.html) so it very much depends on whether or not the used matplotlib backend already supports it (at the moment `QT5Agg` and `tkinter` work nicely, but web-based backends such as `webagg` or `ipympl` still lack full blitting-support so some of the functionalities might not work smoothly as they could... 

If you can provide me with a (ideally very minimalistic) code-snippet on how to embed a basic matplotlib plot I can adapt it to show you how you would use it with EOmaps.

**Comment by raphaelquast** [comment, marked as answer]
@DomagojHack I converted this issue to a discussion to preserve it for other users as well.

I did a quick check using this [matplotlib example](https://matplotlib.org/stable/gallery/user_interfaces/embedding_in_wx2_sgskip.html)... 

At a first glance all seems to work nicely out of the box! 🥳 

<details>

(I used a fresh environment created with `mamba create -n wx -c conda-forge eomaps wxpython`)

<summary> 🐍 source-code </summary>

```python
from matplotlib.backends.backend_wxagg import (
    FigureCanvasWxAgg as FigureCanvas,
    NavigationToolbar2WxAgg as NavigationToolbar)
from matplotlib.figure import Figure

import numpy as np

import wx
import wx.lib.mixins.inspection as WIT

from eomaps import Maps

class CanvasFrame(wx.Frame):
    def __init__(self):
        super().__init__(None, -1, 'EOmaps in a wxpython canvas', size=(550, 350))

        # -------- create EOmaps map ----------
        m = Maps(Maps.CRS.Orthographic(), layer="circles")
        m.all.add_feature.preset.coastline()

        m.cb.click.attach.annotate(zorder=99)
        m.add_feature.preset.ocean()

        m2 = m.new_layer("rectangles")
        m2.cb.click.attach.annotate(zorder=99)
        m2.add_feature.preset.ocean(fc=".25")

        n_indicators = 5
        colors = "rgbcm"
        for i, (r, c) in enumerate(zip(np.logspace(4, 9, n_indicators), colors)):
            m.cb.click.attach.mark(permanent=False, shape="geod_circles",
                                   n=100, lw=2, radius=r, fc="none",
                                   ec=c)

            m2.cb.click.attach.mark(permanent=False, shape="geod_circles",
                                    n=5, lw=2, radius=r, fc="none",
                                    ec=c)

        m.util.layer_selector(loc="upper left", layers=["circles", "rectangles"])
        m.subplots_adjust(top=.9, bottom=0.1, left=0, right=1)

        # set figure and axes
        self.figure = m.figure.f
        self.axes = m.ax
        # ------------------------------------

        self.canvas = FigureCanvas(self, -1, self.figure)

        self.sizer = wx.BoxSizer(wx.VERTICAL)
        self.sizer.Add(self.canvas, 1, wx.LEFT | wx.TOP | wx.EXPAND)
        self.SetSizer(self.sizer)
        self.Fit()

        self.add_toolbar()  # comment this out for no toolbar

    def add_toolbar(self):
        self.toolbar = NavigationToolbar(self.canvas)
        self.toolbar.Realize()
        # By adding toolbar in sizer, we are able to put it at the bottom
        # of the frame - so appearance is closer to GTK version.
        self.sizer.Add(self.toolbar, 0, wx.LEFT | wx.EXPAND)
        # update the axes menu on the toolbar
        self.toolbar.update()


# Alternatively you could use:
# class App(wx.App):
class App(WIT.InspectableApp):
    def OnInit(self):
        """Create the main window and insert the custom frame."""
        self.Init()
        frame = CanvasFrame()
        frame.Show(True)

        return True


if __name__ == "__main__":
    app = App()
    app.MainLoop()
```


</details>

<img src=https://user-images.githubusercontent.com/22773387/172069166-064277a7-9002-4696-85e1-abf9a646e654.gif width=60%>

**Reply by domagojhack** [reply]
Thank you for reply. I will probably use EOmaps for my wx app project and keep you posted if something is not working.

Cheers.

**Reply by raphaelquast** [reply]
Sure! If you end up using it and all works as expected, I'd love to get a look at what you did as well 😅

**Reply by domagojhack** [reply]
I installed EOmaps using conda and using mamba but I can't start the example you provided.

I am getting the following error:
```
Traceback (most recent call last):
  File "/home/domagoj/dev/cadapti/test.py", line 72, in OnInit
    frame = CanvasFrame()
  File "/home/domagoj/dev/cadapti/test.py", line 19, in __init__
    m.all.add_feature.preset.coastline()
AttributeError: 'Maps' object has no attribute 'all'
```
I am not sure what I am doing wrong.

**Reply by domagojhack** [reply]
I found what the issue was. Mamba/Conda installed v3 and not v4 of EOmaps. With a fresh environment, the issue was resolved. I hope my other dependencies will not downgrade EOmaps in the future

**Reply by raphaelquast** [reply]
hey, great! I just wanted to write that this very much looks like you have outdated version of EOmaps...
Actually there shouldn't be dependency differences between v3 and  v4... (not sure why you ended up with v3)

**Reply by domagojhack** [reply]
Probably  I had old version of `gdal` or `proj` packages so mamba decided for some reason not to update them. The example is working like a charm.
Cheers.
